src/inspector_agent.cc

```diff
@@ -0,0 +1,517 @@
+                           std::vector<std::string>* output) {
+  uv_mutex_lock(&queue_lock_);
+  (this->*queue).swap(*output);
+  uv_mutex_unlock(&queue_lock_);
+}
+
+// static
+void Agent::WriteCbIO(uv_async_t* async) {
+  Agent* agent = static_cast<Agent*>(async->data);
+  inspector_socket_t* socket = agent->client_socket_;
+  if (socket) {
+    std::vector<std::string> outgoing_messages;
+    agent->SwapBehindLock(&Agent::outgoing_message_queue_, &outgoing_messages);
+    for (auto const& message : outgoing_messages)
+      inspector_write(socket, message.c_str(), message.length());
+  }
+}
+
+void Agent::WorkerRunIO() {
+  sockaddr_in addr;
+  uv_tcp_t server;
+  int err = uv_async_init(&child_loop_, &io_thread_req_, Agent::WriteCbIO);
+  CHECK_EQ(0, err);
+  io_thread_req_.data = this;
+  uv_tcp_init(&child_loop_, &server);
+  uv_ip4_addr("0.0.0.0", port_, &addr);
+  server.data = this;
+  err = uv_tcp_bind(&server,
+                    reinterpret_cast<const struct sockaddr*>(&addr), 0);
+  if (err == 0) {
+    err = uv_listen(reinterpret_cast<uv_stream_t*>(&server), 1,
+                    OnSocketConnectionIO);
+  }
+  if (err == 0) {
+    PrintDebuggerReadyMessage(port_);
+  } else {
+    fprintf(stderr, "Unable to open devtools socket: %s\n", uv_strerror(err));
+    ABORT();
+  }
+  if (!wait_) {
+    uv_sem_post(&start_sem_);
+  }
+  uv_run(&child_loop_, UV_RUN_DEFAULT);
+  uv_close(reinterpret_cast<uv_handle_t*>(&io_thread_req_), nullptr);
+  uv_close(reinterpret_cast<uv_handle_t*>(&server), nullptr);
+  uv_run(&child_loop_, UV_RUN_DEFAULT);
+}
+
+void Agent::OnInspectorConnectionIO(inspector_socket_t* socket) {
+  if (client_socket_) {
+    return;
+  }
+  client_socket_ = socket;
+  inspector_read_start(socket, OnBufferAlloc, Agent::OnRemoteDataIO);
+  platform_->CallOnForegroundThread(parent_env_->isolate(),
+      new SetConnectedTask(this, true));
+}
+
+void Agent::PostMessages() {
+  if (dispatching_messages_)
+    return;
+  dispatching_messages_ = true;
+  std::vector<std::string> messages;
+  SwapBehindLock(&Agent::message_queue_, &messages);
+  for (auto const& message : messages)
+    inspector_->dispatchMessageFromFrontend(
+        String16::fromUTF8(message.c_str(), message.length()));
+  uv_async_send(&data_written_);
+  dispatching_messages_ = false;
+}
+
+void Agent::SetConnected(bool connected) {
+  if (connected_ == connected)
+    return;
+
+  connected_ = connected;
+  if (connected) {
+    fprintf(stderr, "Debugger attached.\n");
+    inspector_->connectFrontend(new ChannelImpl(this));
+  } else {
+    if (!shutting_down_)
+      PrintDebuggerReadyMessage(port_);
+    inspector_->quitMessageLoopOnPause();
+    inspector_->disconnectFrontend();
+  }
+}
+
+void Agent::Write(const std::string& message) {
+  PushPendingMessage(&outgoing_message_queue_, message);
+  ASSERT_EQ(0, uv_async_send(&io_thread_req_));
+}
+}  // namespace debugger
+}  // namespace node
```

Comment:
The `reinterpret_cast`s in `inspector_socket.cc` are currently necessary. Example:
```
../src/inspector_socket.cc:661:38: error: static_cast from 'uv_tcp_t *' (aka 'uv_tcp_s *') to 'uv_handle_t *' (aka 'uv_handle_s *'), which are not related by inheritance, is not allowed
```
With some C++ template use we may be able to avoid these. Unless you have strong objections, this refactoring can happen later™.